pkgs/top-level/all-packages.nix

```diff
@@ -111,6 +111,16 @@ with pkgs;
 
   nix-update-script = callPackage ../common-updater/nix-update.nix { };
 
+  ### Push NixOS images into the fixed point
```

Comment:
Are they really _inside_ the fixpoint or dangling from it?
They don't use any other value from it than `system`. `pkgsStatic.nixosImages` is wrong; so are the cross sets.
Part of the problem of course is the `nixpkgs.nix` module, which is somewhat mandatory because of its existence in `module-list.nix`; mkForcing `_module.args.pkgs` or settings `nixpkgs.pkgs` is counter to traditional NixOS expectations, but probably ok if you use `assertions` to make sure no `nixpkgs.overlays` or `nixpkgs.config` are set.
If we insert `pkgs` into the images, which is not a technical problem, we do get a little eval performance improvement, as we evaluate `pkgs` once for all images instead of over and over for each.

Reply:
I don't think restricting is necessary, although it does reduce the changes to `release.nix` to make this work. The conditional attr does make the fixpoint attrset strict in `stdenv`, so that could slow down exclusive use of `pkgsStatic` for example (if it isn't already slowed down by that), but that can be fixed by supporting cross invocations later.

First, what you can do is define a module here that you pass as a new optional argument to `release.nix`, which in turn adds it directly in the list of modules that `makeModules` returns, and to the other `modules =` params.
The module then only has to set `nixpkgs.pkgs` and make sure no `nixpkgs.config` is set. Overlays are actually already supported by the `nixpkgs.nix` module.